### PR TITLE
HAAR-879 Amend prison user email

### PR DESCRIPTION
**Pull request by AlTurner-MOJ**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/AuthApiService.kt [hunk 1]
```diff
@@ -70,6 +70,12 @@ class AuthApiService(
       ),
     )
 
+  fun recognised(username: String) =
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/AuthApiService.kt:73]: better name might be usreRecognised - but not a biggie

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt [hunk 2]
```diff
@@ -31,7 +33,62 @@ import javax.validation.constraints.NotEmpty
...
+    ],
+  )
+  fun amendUserEmail(
+    @Parameter(description = "The username of the user.", required = true) @PathVariable
+    username: String,
+    @Valid @RequestBody
+    amendEmail: AmendEmail,
+  ): String? {
+    val link = nomisUserService.changeEmail(username, amendEmail.email!!)
+    return if (smokeTestEnabled) link else ""
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt:89]: Still not sure if the smoketest will call out to here.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/EntityNotFoundException.kt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+package uk.gov.justice.digital.hmpps.manageusersapi.service
+
+class EntityNotFoundException(message: String) :
+  Exception(message)
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/EntityNotFoundException.kt:4]: This is just the same as the NotFoundException.  We should just use one

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt [hunk 4]
```diff
@@ -31,7 +33,62 @@ import javax.validation.constraints.NotEmpty
 @Validated
 class UserController(
   private val nomisUserService: UserService,
+  @Value("\${application.smoketest.enabled}") private val smokeTestEnabled: Boolean,
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt:36]: Still not sure we need this. Does the smoketest ever call manage users?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt [hunk 5]
```diff
@@ -31,7 +33,62 @@ import javax.validation.constraints.NotEmpty
...
+        description = "OK",
+      ),
+      ApiResponse(
+        responseCode = "400",
+        description = "Bad request e.g. missing email address.",
+        content = [
+          Content(
+            mediaType = "application/json",
+            schema = Schema(implementation = ErrorResponse::class),
+          ),
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt:57]: Was just copied from Auth - have we raised a swagger consistency ticket?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/AuthApiService.kt (hunk 1 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/AuthApiService.kt:69]: Happy to change, but kind of like it being succinct where the usage context makes it obvious - felt like it would in this case as we're only passing in the user name. Happy to change though

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt (hunk 4 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt:36]: I don't actually know - this property is used in a similar way on the other amend user email end point, so assumed this would also be needed. Where would I need to look to confirm? Maybe best to cover both of those off with a separate ticket to check it and either leave or remove them?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/AuthApiService.kt (hunk 1 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/AuthApiService.kt]: This is now called confirmRecognised - please let me know what you think

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt (hunk 4 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt:36]: Raising separate ticket to cover this

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserControllerIntTest.kt [hunk 6]
```diff
@@ -15,6 +15,70 @@ import uk.gov.justice.digital.hmpps.manageusersapi.integration.IntegrationTestBa
...
+
+      webTestClient.post().uri("/prisonusers/$username/email")
+        .headers(setAuthorisation(roles = listOf("ROLE_MAINTAIN_ACCESS_ROLES_ADMIN")))
+        .body(
+          fromValue(
+            mapOf("email" to newEmailAddress),
+          ),
+        )
+        .exchange()
+        .expectStatus().isOk
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserControllerIntTest.kt:78]: Do we need tests to ensure we fail gracefully for failures - also - can an empty link ever be returned when not in a smoke test

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt [hunk 7]
```diff
@@ -30,5 +31,26 @@ class UserFixture {
       lastLoggedIn = lastLoggedIn,
       inactiveReason = inactiveReason,
     )
+
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt:34]: Thes prob don't need to be in a companion object - they are just convenience methods on UserFixture

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/prison/UserServiceTest.kt [hunk 8]
```diff
@@ -6,85 +6,164 @@ import org.junit.jupiter.api.Nested
...
+
+      verify(verifyEmailService).requestVerification(prisonUser, newEmailAddress)
+    }
+
+    @Test
+    fun `should update email address in Auth`() {
+      val prisonUser = createNomisUserDetails()
+      whenever(prisonUserApiService.findUserByUsername(userName)).thenReturn(prisonUser)
+      whenever(verifyEmailService.requestVerification(prisonUser, newEmailAddress)).thenReturn(
+        VerifyEmailService.LinkEmailAndUsername("link", newEmailAddress, userName),
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/prison/UserServiceTest.kt:94]: can this be imported?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt [hunk 9]
```diff
@@ -219,3 +276,9 @@ data class NewPrisonUserDto(
     }
   }
 }
+
+data class AmendEmail(
+  @Schema(required = true, description = "Email address", example = "nomis.user@someagency.justice.gov.uk")
+  @field:NotBlank(message = "Email must not be blank")
+  val email: String?,
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserController.kt:283]: It is optional in Auth (although the field has a `NotBlank` validation annotation). Will change as not sure it makes sense to be optional, unless this end point is being used to effectively delete an existing email address?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/prison/UserService.kt [hunk 10]
```diff
@@ -11,32 +11,46 @@ import uk.gov.justice.digital.hmpps.manageusersapi.resource.prison.CreateUserReq
...
+      DPS_GEN -> prisonUserApiService.createGeneralUser(user)
+      DPS_LSA -> prisonUserApiService.createLocalAdminUser(user)
     }
     notificationService.newPrisonUserNotification(user, "DPSUserCreate")
     return nomisUserDetails
   }
 
   fun findUsersByFirstAndLastName(firstName: String, lastName: String): List<PrisonUser> {
-    val nomisUsers: List<PrisonUserSummary> = nomisUserApiService.findUsersByFirstAndLastName(firstName, lastName)
+    val nomisUsers: List<PrisonUserSummary> = prisonUserApiService.findUsersByFirstAndLastName(firstName, lastName)
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/prison/UserService.kt]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt (hunk 7 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/fixtures/UserFixture.kt:34]: Discussed & agreed to leave as is for now

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserControllerIntTest.kt (hunk 6 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/prison/UserControllerIntTest.kt:78]: Discussed and have a added integration test case to ensure that remote failure passed on to client.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/prison/UserServiceTest.kt (hunk 8 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/prison/UserServiceTest.kt]: Had been defined as a nested data class (rightly or wrongly), so needed to be qualified with containing class name. I've made it (plus another exception) top level classes (defined in same file)

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/prison/UserServiceTest.kt [hunk 11]
```diff
@@ -6,85 +6,164 @@ import org.junit.jupiter.api.Nested
...
+      )
+
+      prisonUserService.changeEmail(userName, newEmailAddress)
+
+      verify(authApiService).updateEmail(userName, newEmailAddress)
+    }
+
+    @Test
+    fun `should respond with verify link`() {
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/prison/UserServiceTest.kt:104]: Email validation is covered in VerifyEmailServiceTest

**Review by karenmillermoj — APPROVED** [review]
lgtm
